From: Debian DNS Packaging <pkg-dns-devel@lists.alioth.debian.org>
Date: Fri, 24 Nov 2017 16:26:55 +0000
Subject: Add_--install-layout=deb_to_setup.py_call

---
 bin/python/Makefile.in | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/bin/python/Makefile.in b/bin/python/Makefile.in
index dc1af40..a87927a 100644
--- a/bin/python/Makefile.in
+++ b/bin/python/Makefile.in
@@ -45,9 +45,9 @@ install:: ${TARGETS} installdirs
 	${INSTALL_SCRIPT} dnssec-keymgr ${DESTDIR}${sbindir}
 	if test -n "${PYTHON}" ; then \
 		if test -n "${DESTDIR}" ; then \
-			${PYTHON} ${srcdir}/setup.py install --root=${DESTDIR} --prefix=${prefix} @PYTHON_INSTALL_LIB@ ; \
+			${PYTHON} ${srcdir}/setup.py install --root=${DESTDIR} --prefix=${prefix} @PYTHON_INSTALL_LIB@ --install-layout=deb ; \
 		else \
-			${PYTHON} ${srcdir}/setup.py install --prefix=${prefix} @PYTHON_INSTALL_LIB@ ; \
+			${PYTHON} ${srcdir}/setup.py install --prefix=${prefix} @PYTHON_INSTALL_LIB@ --install-layout=deb ; \
 		fi ; \
 		rm -rf build ; \
 	fi
